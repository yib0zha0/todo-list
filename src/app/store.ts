import { configureStore } from '@reduxjs/toolkit';
import todoListReducer from '../features/TodoList/todoListSlice';

export default configureStore({
    reducer: {
        todoList: todoListReducer,
    }
});

export const buildStore = () => configureStore({
    reducer: {
        todoList: todoListReducer,
    }
});
