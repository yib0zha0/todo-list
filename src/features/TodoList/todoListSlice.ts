import { Todo } from '../TodoEntry/TodoEntry';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export interface TodoListState {
    sortByCreationTime: boolean;
    sortByPriority: boolean;
    todoInProgress: Todo[];
}

export const initialState: TodoListState = {
    sortByCreationTime: true,
    sortByPriority: false,
    todoInProgress: [],
}

export interface TodoModifyPayload {
    todoIndex: number,
    field: string,
    value: any,
}

const todoListSlice = createSlice({
    name: 'todoList',
    initialState: initialState,
    reducers: {
        addTodo: (state) => {
            const newTodo = {
                id: `todo-0`, 
                completed: false, 
                title: '', 
                description: '', 
                priority: 1, 
                creationTime: new Date().getTime(),
            };
            state.todoInProgress.push(newTodo);
        },
        deleteTodo: (state, action: PayloadAction<number>) => {
            const newTodos = state.todoInProgress.filter((value, index) => index != action.payload);
            state.todoInProgress = newTodos;
        },
        modifyTodo: (state, action: PayloadAction<TodoModifyPayload>) => {
            const todo = state.todoInProgress[action.payload.todoIndex];
            const field = action.payload.field;
            if ('completed' === field) {
                todo.completed = action.payload.value;
            } else if ('title' === field) {
                todo.title = action.payload.value;
            } else if ('description' === field) {
                todo.description = action.payload.value;
            } else if ('priority' === field) {
                todo.priority = action.payload.value;
            }
        },
        sortByCreationTime: (state) => {
            state.sortByCreationTime = true;
            state.sortByPriority = false;
            state.todoInProgress.sort((todo1, todo2) => todo1.creationTime - todo2.creationTime);
        },
        sortByPriority: (state) => {
            state.sortByCreationTime = false;
            state.sortByPriority = true;
            state.todoInProgress.sort((todo1, todo2) => todo1.priority - todo2.priority);
        }
    },
});

export const { addTodo, deleteTodo, modifyTodo, sortByCreationTime, sortByPriority } = todoListSlice.actions;

export const todoListSelector = (state: {todoList: TodoListState}) => state.todoList;

export default todoListSlice.reducer;