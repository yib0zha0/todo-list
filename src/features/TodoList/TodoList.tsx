import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { todoListSelector } from './todoListSlice';
import { View, Button, StyleSheet, Text } from 'react-native';
import { addTodo, sortByPriority, sortByCreationTime } from './todoListSlice';
import TodoEntry from '../TodoEntry/TodoEntry';

const TodoList = () => {
    const todoList = useSelector(todoListSelector);
    const dispath = useDispatch();

    return (
        <View>
            <View style={styles.sortButtonGroup}>
                <Text>Sort By: </Text>
                <View style={styles.sortButton}>
                    <Button 
                        testID={'sortByPriority'}
                        title={'Priority'}
                        onPress={(event) => dispath(sortByPriority())}
                    />
                </View>
                <View style={styles.sortButton}>
                    <Button 
                        testID={'sortByTime'}
                        title={'Time'}
                        onPress={(event) => dispath(sortByCreationTime())}
                    />
                </View>
            </View>
            {
                todoList.todoInProgress.map((todo, index) => <TodoEntry todo={todo} index={index} key={index}/>)
            }
            <Button 
                testID={'newTodo'}
                title={'New Todo'}
                onPress={(event) => dispath(addTodo())}
            />
        </View>
    );
}

const styles = StyleSheet.create({
    sortButtonGroup: {
        flexDirection: 'row',
        paddingVertical: 2,
    },
    sortButton: {
        flex: 0.3,
        paddingHorizontal: 2,
    }
});

export default TodoList;