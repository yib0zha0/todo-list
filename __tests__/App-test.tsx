/**
 * @format
 */

import 'react-native';
import React from 'react';
import App from '../src/App';
import { initialState } from '../src/features/TodoList/todoListSlice';
import { buildStore } from '../src/app/store';
import todoListReducer from '../src/features/TodoList/todoListSlice';
import { Provider } from 'react-redux';
import { Button, Picker } from 'react-native';
import { render, cleanup, fireEvent } from '@testing-library/react-native';

afterEach(cleanup);

describe('', () => {

  it('snapshot test', () => {
    const store = buildStore();
    const rendered = render(<Provider store={store}><App /></Provider>);
    const newTodoButton = rendered.getByTestId('newTodo');
    fireEvent.press(newTodoButton);
    const snapshot = rendered.toJSON();
    expect(snapshot).toMatchSnapshot();
    console.log()
  });

  it('add todo', () => {
    const store = buildStore();
    const rendered = render(<Provider store={store}><App /></Provider>);
    const newTodoButton = rendered.getByTestId('newTodo');
    fireEvent.press(newTodoButton);
    const todoList = store.getState().todoList;
    console.log(store.getState());
    expect(todoList.todoInProgress).toHaveLength(1);
    const todoEntry = todoList.todoInProgress[0];
    expect(todoEntry.id).toBeDefined();
    expect(todoEntry.completed).toBeFalsy();
    expect(todoEntry.priority).toBe(1);
    expect(todoEntry.creationTime).toBeDefined();
    expect(todoEntry.title).toHaveLength(0);
    expect(todoEntry.description).toHaveLength(0);
  })

  it('edit todo', () => {
    const store = buildStore();
    const rendered = render(<Provider store={store}><App /></Provider>);
    const title = 'New title';
    const description = 'New description';

    const newTodoButton = rendered.getByTestId('newTodo');
    fireEvent.press(newTodoButton);
    const titleInput = rendered.getByPlaceholderText('Title');
    fireEvent.changeText(titleInput, title);
    const descriptionInput = rendered.getByPlaceholderText('Description');
    fireEvent.changeText(descriptionInput, description);
    const checkbox = rendered.getByTestId('checkbox-todo-0');

    const todoList = store.getState().todoList;
    expect(todoList.todoInProgress).toHaveLength(1);
    const todoEntry = todoList.todoInProgress[0];
    expect(todoEntry.title).toBe(title);
    expect(todoEntry.description).toBe(description);
    expect(todoEntry.completed).toBeFalsy();
  })

  it('delete todo', () => {
    const store = buildStore();
    const rendered = render(<Provider store={store}><App /></Provider>);

    const newTodoButton = rendered.getByTestId('newTodo');
    fireEvent.press(newTodoButton);
    expect(store.getState().todoList.todoInProgress).toHaveLength(1);

    const deleteButton = rendered.getByTestId('delete-todo-0');
    fireEvent.press(deleteButton);
    expect(store.getState().todoList.todoInProgress).toHaveLength(0);
  });
});


