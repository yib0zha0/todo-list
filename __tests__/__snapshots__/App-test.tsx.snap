// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
  >
    <RCTScrollView
      contentInsetAdjustmentBehavior="automatic"
      style={
        Object {
          "backgroundColor": "#F3F3F3",
        }
      }
    >
      <View>
        <View
          accessibilityIgnoresInvertColors={true}
          style={
            Object {
              "backgroundColor": "#F3F3F3",
              "paddingBottom": 40,
              "paddingHorizontal": 32,
              "paddingTop": 96,
            }
          }
        >
          <Image
            accessibilityRole="image"
            source={
              Object {
                "testUri": "../Libraries/NewAppScreen/components/logo.png",
              }
            }
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {
                  "height": undefined,
                  "width": undefined,
                },
                Object {
                  "marginBottom": -192,
                  "marginLeft": -128,
                  "opacity": 0.2,
                  "overflow": "visible",
                  "resizeMode": "cover",
                },
              ]
            }
          />
          <Text
            style={
              Object {
                "color": "#000",
                "fontSize": 40,
                "fontWeight": "600",
                "textAlign": "center",
              }
            }
          >
            Welcome to React
          </Text>
        </View>
        <View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "paddingVertical": 2,
              }
            }
          >
            <Text>
              Sort By: 
            </Text>
            <View
              style={
                Object {
                  "flex": 0.3,
                  "paddingHorizontal": 2,
                }
              }
            >
              <View
                accessibilityRole="button"
                accessibilityState={Object {}}
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="sortByPriority"
              >
                <View
                  style={
                    Array [
                      Object {},
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#007AFF",
                          "fontSize": 18,
                          "margin": 8,
                          "textAlign": "center",
                        },
                      ]
                    }
                  >
                    Priority
                  </Text>
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "flex": 0.3,
                  "paddingHorizontal": 2,
                }
              }
            >
              <View
                accessibilityRole="button"
                accessibilityState={Object {}}
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="sortByTime"
              >
                <View
                  style={
                    Array [
                      Object {},
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#007AFF",
                          "fontSize": 18,
                          "margin": 8,
                          "textAlign": "center",
                        },
                      ]
                    }
                  >
                    Time
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "floralwhite",
                "borderColor": "lavender",
                "borderWidth": 2,
                "padding": 5,
              }
            }
            testID="todo-0"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "borderBottomColor": "lavender",
                  "borderBottomWidth": 2,
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <View
                pointerEvents="auto"
              >
                <RNCCheckbox
                  forwardedRef={null}
                  onValueChange={[Function]}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "height": 32,
                        "width": 32,
                      },
                      Object {
                        "flex": 0.1,
                      },
                    ]
                  }
                  testID="checkbox-todo-0"
                  value={false}
                />
              </View>
              <TextInput
                allowFontScaling={true}
                onChangeText={[Function]}
                placeholder="Title"
                rejectResponderTermination={true}
                style={
                  Object {
                    "flex": 0.6,
                    "fontSize": 20,
                  }
                }
                underlineColorAndroid="transparent"
                value=""
              />
              <View
                style={
                  Object {
                    "flex": 0.3,
                  }
                }
              >
                <View>
                  <RCTPicker
                    items={
                      Array [
                        Object {
                          "label": "Urgent",
                          "textColor": undefined,
                          "value": "1",
                        },
                        Object {
                          "label": "High",
                          "textColor": undefined,
                          "value": "2",
                        },
                        Object {
                          "label": "Medium",
                          "textColor": undefined,
                          "value": "3",
                        },
                        Object {
                          "label": "Low",
                          "textColor": undefined,
                          "value": "4",
                        },
                      ]
                    }
                    onChange={[Function]}
                    selectedIndex={0}
                    style={
                      Array [
                        Object {
                          "height": 216,
                        },
                        undefined,
                      ]
                    }
                  />
                </View>
              </View>
              <View
                style={
                  Object {
                    "flex": 0.1,
                  }
                }
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={Object {}}
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "opacity": 1,
                    }
                  }
                  testID="delete-todo-0"
                >
                  <View
                    style={
                      Array [
                        Object {},
                      ]
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#007AFF",
                            "fontSize": 18,
                            "margin": 8,
                            "textAlign": "center",
                          },
                          Object {
                            "color": "red",
                          },
                        ]
                      }
                    >
                      X
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <TextInput
              allowFontScaling={true}
              onChangeText={[Function]}
              placeholder="Description"
              rejectResponderTermination={true}
              style={Object {}}
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <View
            accessibilityRole="button"
            accessibilityState={Object {}}
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="newTodo"
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                New Todo
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
</View>
`;
